outlines/types/dsl.py

```diff
@@ -0,0 +1,379 @@
+
+
+@dataclass
+class QuantifyExact(Term):
+    term: Term
+    count: int
+
+    def _display_node(self) -> str:
+        return f"Quantify({{{self.count}}})"
+
+    def _display_children(self, indent: str) -> str:
+        return self.term.display_ascii_tree(indent, True)
+
+    def __repr__(self):
+        return f"QuantifyExact(term={repr(self.term)}, count={repr(self.count)})"
+
+
+@dataclass
+class QuantifyMinimum(Term):
+    term: Term
+    min_count: int
+
+    def _display_node(self) -> str:
+        return f"Quantify({{{self.min_count},}})"
+
+    def _display_children(self, indent: str) -> str:
+        return self.term.display_ascii_tree(indent, True)
+
+    def __repr__(self):
+        return (
+            f"QuantifyMinimum(term={repr(self.term)}, min_count={repr(self.min_count)})"
+        )
+
+
+@dataclass
+class QuantifyMaximum(Term):
+    term: Term
+    max_count: int
+
+    def _display_node(self) -> str:
+        return f"Quantify({{,{self.max_count}}})"
+
+    def _display_children(self, indent: str) -> str:
+        return self.term.display_ascii_tree(indent, True)
+
+    def __repr__(self):
+        return (
+            f"QuantifyMaximum(term={repr(self.term)}, max_count={repr(self.max_count)})"
+        )
+
+
+@dataclass
+class QuantifyBetween(Term):
+    term: Term
+    min_count: int
+    max_count: int
+
+    def __post_init__(self):
+        if self.min_count > self.max_count:
+            raise ValueError(
+                "QuantifyBetween: `max_count` must be greater than `min_count`."
+            )
+
+    def _display_node(self) -> str:
+        return f"Quantify({{{self.min_count},{self.max_count}}})"
+
+    def _display_children(self, indent: str) -> str:
+        return self.term.display_ascii_tree(indent, True)
+
+    def __repr__(self):
+        return f"QuantifyBetween(term={repr(self.term)}, min_count={repr(self.min_count)}, max_count={repr(self.max_count)})"
+
+
+def repeat(self: Term, min_count: int, max_count: int) -> QuantifyBetween:
+    match (min_count, max_count):
+        case (None, None):
+            raise ValueError(
+                "repeat: you must provide a value for at least `min_count` or `max_count`"
+            )
+        case (_, None):
+            return QuantifyMinimum(self, min_count)
+        case (None, _):
+            return QuantifyMaximum(self, max_count)
+        case _:
+            return QuantifyBetween(self, min_count, max_count)
+
+
+def times(self: Term, count: int = 0) -> QuantifyExact:
+    return QuantifyExact(self, count)
+
+
+def optional(self: Term) -> Optional:
+    return Optional(self)
+
+
+def one_or_more(self: Term) -> KleenePlus:
+    return KleenePlus(self)
+
+
+def zero_or_more(self: Term) -> KleeneStar:
+    return KleeneStar(self)
+
+
+Term.one_or_more = one_or_more
+Term.optional = optional
+Term.repeat = repeat
+Term.times = times
+Term.zero_or_more = zero_or_more
+
+
+def to_regex(term: Term) -> str:
```

Comment:
Just want to draw attention to a few terms that will need extra sets of parentheses:

```python
>>> to_regex(QuantifyExact(String("test"), 3))
'test{3}'
>>> to_regex(Sequence([Regex("dog|cat"), String("fish")]))
'dog|catfish'
```

I realise this PR is still in draft though :)